### ISSUE TITLE
Error 11 when running arks + links pipeline

### ISSUE THREAD

**Issue by Xuyue-Yang** [bcgsc/arcs#77]
Hello,

I am running Arks + links pipeline on a Supernova assembly.
`arks-make arks draft=draft reads=reads threads=10 time=1 o=3`

 And got error message like:
```
command time -v -o sample_c5_m50-10000_k30_r0.05_e30000_z500_l5_a0.3.scaffolds.fa.time LINKS -f sample.renamed.fa -s empty.fof -b sample_c5_m50 10000_k30_r0.05_e30000_z500_l5_a0.3 -l 5 -a 0.3 -z 500 
rm sample_c5_m50-10000_k30_r0.05_e30000_z500_l5_a0.3.tigpair_checkpoint.tsv
make: *** [sample_c5_m50-10000_k30_r0.05_e30000_z500_l5_a0.3.scaffolds.fa] Error 11
```

It used to work fine when I ran the same code with ./arks-make arks-tigmint but now reports error 11 without further information. 

Your help is greatly appreciated,
Many thanks.

**Comment by lcoombe** [comment]
Hi @Xuyue-Yang,

I see that the LINKS step was executed -- did that complete successfully? You could take a look at the log from LINKS to see whether there was an error in that step.

Are you running the pipeline on the same machine as your previous runs?

**Comment by Xuyue-Yang** [comment]
Hi,
Yes I was running the pipeline on the exact same machine. These are the output I got so far before the program stopped. 

```
sample_c5_m50-10000_k30_r0.05_e30000_z500_contigrec.tsv
sample_c5_m50-10000_k30_r0.05_e30000_z500_imap.tsv
sample_c5_m50-10000_k30_r0.05_e30000_z500_kmercontigrec.tsv
sample_c5_m50-10000_k30_r0.05_e30000_z500_l5_a0.3.scaffolds.fa.time
sample_c5_m50-10000_k30_r0.05_e30000_z500_original.gv
sample_c5_m50-10000_k30_r0.05_e30000_z500_original.gv.time
sample_c5_m50-10000_k30_r0.05_e30000_z500.tigpair_checkpoint.tsv
sample_multiplicities.csv
```
And I did not seem to find a log file. Or is the log file located somewhere else?

Best regards,
Xuyue

**Comment by lcoombe** [comment]
Hi @Xuyue-Yang,

Did you pipe the standard out of the commands anywhere else? When we run the pipeline, we get a file with a `.log` suffix from the LINKS step.

Is your LINKS installation being found/run properly from your PATH? (ie. try `LINKS` with no parameters to check) 
Are you using the most up-to-date Makefile? We integrated ARKS into this ARCS repo, so you can run the most current version by using `arcs-make arks`.

Could you post all the logs that you can find (ie. steps before LINKS as well)? That would be very helpful for me to help you figure out what is going wrong. You could also post the full dry run output (ie. run the same makefile command, but specify `-n` to print all commands that will be executed, but doesn't execute them).

**Comment by Xuyue-Yang** [comment]
Hi Lauren,
 I tried to run LINKS with no parameters and it did not report error. Also links works fine when I used `arks-make arks-tigmint` command.
I only tried to install `arks-make` at the first place because it is a small genome and not believed to be very repetitive. Maybe I should also try to install `arcs` and see how it works.

Here I attached all the logs I got. Hope now it is informative.

```
touch empty.fof
echo barcoded.fq.gz > barcoded.fof
/Pathtoarksmake/../bin/calcBarcodeMultiplicities.pl barcoded.fof > barcoded_multiplicities.csv
perl -ne 'chomp; if(/>/){$ct+=1; print ">$ct\n";}else{print "$_\n";} ' < sample.fa > sample.renamed.fa 
command time -v -o sample_c5_m50-10000_k30_r0.05_e30000_z500_original.gv.time arks -p full -v -f sample.renamed.fa -a barcoded_multiplicities.csv -c 5 -t 10 -j 0.55 -o 3 -m 50-10000 -k 30 -r 0.05 -e 30000 -z 500 -d 0 -b sample_c5_m50-10000_k30_r0.05_e30000_z500 barcoded.fq.gz 
Reading user inputs...
Finished reading user inputs...entering runArks()...
Entered runArks()...
Running: arks 1.0.3
 pid 11723
 -p full
 -f sample.renamed.fa
 -a barcoded_multiplicities.csv
 -q 
 -w 
 -i 
 -o 3
 -c 5
 -k 30
 -g 1
 -j 0.55
 -l 0
 -z 500
 -b sample_c5_m50-10000_k30_r0.05_e30000_z500
 Min index multiplicity: 50
 Max index multiplicity: 10000
 -d 0
 -e 30000
 -r 0.05
 -t 10
 -v 1

---We are using KMER method.---


=>Preprocessing: Gathering barcode multiplicity information...Wed Oct 30 19:50:44 2019
Saw 3216606 barcodes and keeping 492496496 read pairs out of 492496496

=>Preprocessing: Gathering draft information...Wed Oct 30 19:50:48 2019


----Full ARKS----


=>Storing Kmers from Contig ends... Wed Oct 30 19:50:49 2019

Finished 1000 Contigs...
......
Finished 11000 Contigs...
Total number of contigs in draft genome:  11686
Total valid contigs:  11686
Total skipped contigs:  0
Total number of Kmers:  146935018
Number Null Kmers:  249767
Number Kmers Recorded:  101913591
Number Kmer Collisions:  45021427
Number Times Kmers Removed (since duplicate in different contig):  44658589
Number of unique kmers (only one contig):  91367638

=>Reading Chromium FASTQ file(s)... Wed Oct 30 19:54:15 2019

Reading chrom barcoded.fq.gz
Processed 10000000 read pairs.
......
Processed 540000000 read pairs.
Stored read pairs: 12993586
Skipped invalid read pairs: 75923
Skipped unpaired reads: 0
Skipped reads pairs without a good contig: 233254662
Total valid kmers: 2859239837
Number invalid kmers: 5236436
Number of kmers found in ContigKmap: 681100318
Number of kmers recorded in Ktrack: 2663591803
Number of kmers found in ContigKmap but duplicate: 2312475811
Number of reads passing jaccard threshold: 36691883
Number of reads failing jaccard threshold: 452338702
WARNING:: Your chromium read file has 26045168 readpairs that have an empty barcode.Cumulative memory usage: 5238932

=>Starting pairing of scaffolds... Wed Oct 30 22:32:35 2019

=>Starting to create graph... Wed Oct 30 22:32:57 2019

=>Starting to write graph file... Wed Oct 30 22:32:58 2019

      Max Degree (-d) set to: 0. Will not delete any verticies from graph.
      Writting graph file to sample_c5_m50-10000_k30_r0.05_e30000_z500_original.gv...

=>Outputting desired checkpoint files... Wed Oct 30 22:32:59 2019


=>Done. Wed Oct 30 22:33:28 2019

python /Pathtoarksmake/../Examples/makeTSVfile.py sample_c5_m50-10000_k30_r0.05_e30000_z500_original.gv sample_c5_m50-10000_k30_r0.05_e30000_z500.tigpair_checkpoint.tsv sample.renamed.fa 
ln -s sample_c5_m50-10000_k30_r0.05_e30000_z500.tigpair_checkpoint.tsv sample_c5_m50-10000_k30_r0.05_e30000_z500_l5_a0.3.tigpair_checkpoint.tsv
command time -v -o sample_c5_m50-10000_k30_r0.05_e30000_z500_l5_a0.3.scaffolds.fa.time LINKS -f sample.renamed.fa -s empty.fof -b sample_c5_m50-10000_k30_r0.05_e30000_z500_l5_a0.3 -l 5 -a 0.3 -z 500 
rm sample_c5_m50-10000_k30_r0.05_e30000_z500_l5_a0.3.tigpair_checkpoint.tsv
Number of contigs:11686
Size of Contig Array:23373
File barcoded.fq.gz opened.
make: *** [sample_c5_m50-10000_k30_r0.05_e30000_z500_l5_a0.3.scaffolds.fa] Error 11


```

**Comment by lcoombe** [comment]
Thanks @Xuyue-Yang ,

This is a bit of a cryptic issue for me -- we've never seen anything like this before. Were your parameters for ARCS/LINKS the same when you ran `arks-tigmint`? There shouldn't be any difference in the success of the executions, as both of those runs would end up running the same LINKS command, just with different input scaffolds files.

Could you try running without `o=3`? We don't normally output the checkpoint files, so it is possible that those are interfering.

Running with the dry-run (`-n`) like I mentioned before might also be helpful to double check the expected command will be run.
Ie.
```
arks-make -n arks draft=draft reads=reads threads=10
``` 

Does LINKS run properly if you run the same command from the Makefile, but just independently on the command line?
Ie. run:
```
command time -v -o sample_c5_m50-10000_k30_r0.05_e30000_z500_l5_a0.3.scaffolds.fa.time LINKS -f sample.renamed.fa -s empty.fof -b sample_c5_m50-10000_k30_r0.05_e30000_z500_l5_a0.3 -l 5 -a 0.3 -z 500
```

**Comment by Xuyue-Yang** [comment]
Hi Lauren,

I tried the dry-run
`arks-make -n arks draft=draft reads=reads threads=10`
and it the report said:
```
ln -s sample_c5_m50-10000_k30_r0.05_e30000_z500.tigpair_checkpoint.tsv sample_c5_m50-10000_k30_r0.05_e30000_z500_l5_a0.3.tigpair_checkpoint.tsv
LINKS -f sample.renamed.fa -s empty.fof -b sample_c5_m50-10000_k30_r0.05_e30000_z500_l5_a0.3 -l 5 -a 0.3 -z 500 
```
Then I ran this in the command line
`LINKS -f sample.renamed.fa -s empty.fof -b sample_c5_m50-10000_k30_r0.05_e30000_z500_l5_a0.3 -l 5 -a 0.3 -z 500 `

The error said: 
```
=>Reading contig/sequence assembly file : Wed Nov  6 13:29:47 CET 2019
Building a Bloom filter using 15-mers derived from sequences in -f sample.renamed.fa...
*****
Bloom filter specs
elements=389024495
FPR=0.001
size (bits)=5593233792
hash functions=9
*****
Contigs (>= 500 bp) processed k=15:
arks.sh: line 64:  4399 Segmentation fault      (core dumped) LINKS -f sample.renamed.fa -s empty.fof -b sample_c5_m50-10000_k30_r0.05_e30000_z500_l5_a0.3 -l 5 -a 0.3 -z 500

```
Then I referred to the core.XXX file

``
[New LWP 4399]
Missing separate debuginfo for the main executable file
Try: yum --enablerepo='*debug*' install /usr/lib/debug/.build-id/d2/5e27c856e53d82edfeb127672e4ec6a021844c
Core was generated by `perl ../links_v1.8.6/LINKS -f sample.'.
Program terminated with signal 11, Segmentation fault.
#0  0x00002ad983e26615 in ?? ()
```

I followed their suggestion by trying (have no idea what I was doing) `yum --enablerepo='*debug*' install /usr/lib/debug/.build-id/d2/5e27c856e53d82edfeb127672e4ec6a021844c
`
And I got 
```
Loaded plugins: nvidia, versionlock
You need to be root to perform this command.
```
I am working in a shared server and have restricted permission. So I stopped here and hopefully you see what the problem is.

Best regards,
Xuyue

**Comment by lcoombe** [comment]
Hi @Xuyue-Yang,

There is something wrong with the LINKS command - It should read from a checkpoint file, not build a Bloom filter (it should only do that step when you are using LINKS for long-read scaffolding, not withing ARCS/ARKS).

This is the example of the message you should see:
```
Scaffolding mid-point files:
test_scaffolds_c5_m50-6000_k30_r0.05_e30000_z500_l5_a0.9.tigpair_checkpoint.tsv
test_scaffolds_c5_m50-6000_k30_r0.05_e30000_z500_l5_a0.9.simplepair_checkpoint.tsv
detected; LINKS will skip contig pairing and use info from files instead: Wed Jun 20 16:21:00 PDT 2018
```

Did you run the `ln -s` command? With that command, LINKS expects to see a contigs file named `sample.renamed.fa` and a checkpoint file named `sample_c5_m50-10000_k30_r0.05_e30000_z500_l5_a0.3.tigpair_checkpoint.tsv`

**Comment by Xuyue-Yang** [comment]
Hi Lauren,

I got the problem fixed by installing the most current ARCS version and ran arks+links by `arcs-make arks`.  I also re-installed the same version of dependencies as your test version in the manual.  The genome got quite a good improvement of N50.

My guess is I did not install arks (or dependencies) properly in the first place. Thank you very much for your effort in this ticket.

Best regards
